### PR TITLE
Clipping ops

### PR DESCRIPTION
**Pull request by rmanoka**
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/main/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

I had a use-case for clipping a 1-d geom. by a 2-d goem.  It's conceptually a bool ops, but needed some refactoring of the region construction implementation.

### CODE REVIEW

File: geo/src/algorithm/bool_ops/op.rs [hunk 1]
```diff
@@ -1,66 +1,69 @@
...
-                is_first,
+                idx,
                 _region: region.into(),
                 _region_2: region.into(),
             });
         }
     }
 
-    pub fn sweep(&self) -> MultiPolygon<T> {
+    // is_first -> whether it is from first input or second input
```
Comment (michaelkirk) [inline geo/src/algorithm/bool_ops/op.rs:54]:
```suggestion
    // idx: whether it is from first input or second input
```

File: geo/src/algorithm/sweep/proc.rs [hunk 2]
```diff
@@ -6,12 +6,13 @@ use std::{
...
 
 pub(crate) struct Sweep<C: Cross> {
+    is_simple: bool,
     events: BinaryHeap<Event<C::Scalar, IMSegment<C>>>,
     active_segments: BTreeSet<Active<IMSegment<C>>>,
 }
 
 impl<C: Cross + Clone> Sweep<C> {
-    pub(crate) fn new<I>(iter: I) -> Self
+    pub(crate) fn new<I>(iter: I, is_simple: bool) -> Self
```
Comment (michaelkirk) [inline geo/src/algorithm/sweep/proc.rs:15]: Is there a definition of what `is_simple` means in this context?

File: geo-bool-ops-benches/benches/boolean_ops.rs [hunk 3]
```diff
@@ -80,7 +83,7 @@ fn run_complex<T: Measurement>(c: &mut Criterion<T>) {
         group.bench_with_input(BenchmarkId::new("geo::intersects", steps), &(), |b, _| {
```
Comment (michaelkirk) [inline geo-bool-ops-benches/benches/boolean_ops.rs:83]:
```suggestion
        group.bench_with_input(BenchmarkId::new("geo::relate", steps), &(), |b, _| {
```

Probably good to keep the output in line with what's being benchmarked - though TBH I'm not sure why we're benchmarking the `intersects` method in the overlay / bool-ops benchmarks.

**Review by michaelkirk — COMMENTED** [review]
I had some clarifying questions and some typo fixes - I'd like to re-review the core of the algorithm once I understand better what is_simple is about.

File: geo/src/algorithm/bool_ops/spec.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+impl<T: GeoFloat> From<OpType> for BoolOp<T> {
+    fn from(ty: OpType) -> Self {
+        Self {
+            ty,
+            assembly: RegionAssembly::default(),
+        }
+    }
+}
+impl<T: GeoFloat> Spec<T> for BoolOp<T> {
+    type Region = Region;
```
[the remark this answers is not in the document]
Comment (rmanoka) [reply geo/src/algorithm/bool_ops/spec.rs:30]: I set that up for flexibility but didn't end up using it.  Looking ahead:  the algo. can actually be generalized to handle "cascade union" / union of a vector of polygons.  In those cases, the region would be different.

File: geo/src/algorithm/bool_ops/tests.rs [hunk 5]
```diff
@@ -161,10 +155,17 @@ fn test_issue_865() -> Result<()> {
...
     let wkt1 = "POLYGON ((20 0, 20 160, 200 160, 200 0, 20 0))";
-    let wkt2 = "POLYGON ((220 80, 0 80, 0 240, 220 240, 220 80),	(100 80, 120 120, 80 120, 100 80))";
+    let wkt2 = "LINESTRING (0 0, 100 100, 240 100)";
 
-    check_sweep(wkt1, wkt2, OpType::Intersection)?;
+    let poly1 = MultiPolygon::<f64>::try_from_wkt_str(wkt1)
+        .or_else(|_| Polygon::<f64>::try_from_wkt_str(wkt1).map(MultiPolygon::from))
+        .unwrap();
+    let mls = MultiLineString::try_from_wkt_str(wkt2)
+        .or_else(|_| LineString::<f64>::try_from_wkt_str(wkt2).map(MultiLineString::from))
```
[the remark this answers is not in the document]
Comment (rmanoka) [reply geo/src/algorithm/bool_ops/tests.rs:166]: Yea, lets merge #882 anyway.  I am not enabling anything more in JTS tests yet here as I haven't looked into the semantics of both.

File: geo/src/algorithm/sweep/iter.rs [hunk 6]
```diff
@@ -115,24 +121,28 @@ where
...
         self.sweep.prev_active(c).map(|s| (s.geom, &s.cross))
     }
-}
 
-impl<C> FromIterator<C> for CrossingsIter<C>
-where
-    C: Cross + Clone,
-{
-    fn from_iter<T: IntoIterator<Item = C>>(iter: T) -> Self {
+    fn new_ex<T: IntoIterator<Item = C>>(iter: T, is_simple: bool) -> Self {
```
[the remark this answers is not in the document]
Comment (rmanoka) [reply geo/src/algorithm/sweep/iter.rs:125]: `ex` is the typical suffix in rust for the "extended" version.  `is_simple` is a flag to make the sweep algorithm skip checking for intersections.  It speeds up our bool ops. but didn't see a great change in benchmarks.

**Review by michaelkirk — APPROVED** [review]
Ok LGTM! Thanks for the code and thanks for helping me review!

**Comment by michaelkirk** [comment]
(reminder that this still needs a changelog entry, but otherwise feel free to merge)

**Comment by michaelkirk** [comment]
Are you waiting on anything else before merging this @rmanoka?

If it's helpful, I can provide the changelog entry if that's the only thing we're waiting on.

**Comment by rmanoka** [comment]
No major changes; didn't get to wrap it up due to travel.  I should get to this by this weekend, but feel free to merge if pushing for a release before then.

**Pull request merged**